src/Microsoft.ML.Tokenizers/PreTokenizer/PreTokenizer.cs

```diff
@@ -54,11 +65,98 @@ public Split(string token, (int Index, int End) offset)
+        /// <param name="skipSpecialTokens">Indicates whether to skip the special tokens.</param>
         /// <returns>The list of the splits containing the tokens and the token's offsets to the original string.</returns>
-        public abstract IReadOnlyList<Split> PreTokenize(string sentence);
+        public abstract IEnumerable<Split> PreTokenize(string sentence, bool skipSpecialTokens = false);
+    }
+
+    internal sealed class RegexSplitEnumerable : IEnumerable<Split>
+    {
+        private readonly static Dictionary<string, Regex> _regexCache = new(StringComparer.Ordinal);
+        private readonly Regex _regex;
+        private readonly string _sentence;
+
+        public RegexSplitEnumerable(string sentence, string pattern)
+        {
+            Debug.Assert(sentence is not null);
+            Debug.Assert(pattern is not null);
+
+            Regex? regex;
+            lock (_regexCache)
+            {
+                if (!_regexCache.TryGetValue(pattern!, out regex))
+                {
+                    regex = new Regex(pattern, RegexOptions.Compiled);
+                    _regexCache[pattern!] = regex;
+                }
+            }
+
+            _regex = regex;
+            _sentence = sentence!;
+        }
+
+        public IEnumerator<Split> GetEnumerator() => new RegexSplitEnumerator(_regex, _sentence);
+
+        IEnumerator IEnumerable.GetEnumerator() => new RegexSplitEnumerator(_regex, _sentence);
+
+        private sealed class RegexSplitEnumerator : IEnumerator<Split>
+        {
+            private Split _current = default;
+            private readonly Regex _regex;
+            private Match? _tokenMatch;
+            private readonly string _sentence;
+
+            public RegexSplitEnumerator(Regex regex, string sentence)
+            {
+                Debug.Assert(sentence is not null);
+                Debug.Assert(regex is not null);
+
+                _regex = regex!;
+                _sentence = sentence!;
+            }
+
+            public Split Current => _current;
+
+            object IEnumerator.Current => _current;
+
+            public bool MoveNext()
+            {
+                if (_tokenMatch is null)
+                {
+                    _tokenMatch = _regex.Match(_sentence);
+                }
+                else if (!_tokenMatch.Success)
+                {
+                    return false;
+                }
+                else
+                {
+                    _tokenMatch = _tokenMatch.NextMatch();
+                }
+
+                if (!_tokenMatch.Success)
+                {
+                    return false;
+                }
+
+                _current = new Split(_tokenMatch.Value, (_tokenMatch.Index, _tokenMatch.Index + _tokenMatch.Length));
+                return true;
+            }
+
+            public void Reset()
+            {
+                _tokenMatch = null;
+            }
+
+            public void Dispose()
```

Comment:
If dispose is going to do nothing is it needed?

Reply:
This is part of the interface we have to implement. Not having it will produce error like:

`error CS0535: 'RegexSplitEnumerable.RegexSplitEnumerator' does not implement interface member 'IDisposable.Dispose()' [C:\oss\machinelearning\src\Microsoft.ML.Tokenizers\Microsoft.ML.Tokenizers.csproj::TargetFramework=netstandard2.0]`